File: 2021/day04.hs
import Data.List
import Data.List.Split

main = do
  (numsStr : boardsStrs) <- splitOn "\n\n" <$> readFile "input/day04"
  let nums = map read $ splitOn "," numsStr :: [Int]
      boards = parseBoard <$> boardsStrs
      (winner : scores) = play nums boards []
  putStrLn $ "Part 1: " ++ show winner
  putStrLn $ "Part 2: " ++ show (last scores)

parseBoard str = let rs = map read . words <$> lines str in (rs, transpose rs)

play _ [] scores = scores
play (num : nums) boards scores = play nums boards' scores'
  where
    (winnerBoards, boards') = partition won $ scratch num <$> boards
    scores' = scores ++ map (score num) winnerBoards

scratch n (rows, cols) = (filter (/= n) <$> rows, filter (/= n) <$> cols)

won (rows, cols) = any null rows || any null cols

score num = (* num) . sum . map sum . fst

File: 2021/day04.py
def parse_board(section):
    rows = [list(map(int, line.split())) for line in section.split('\n')]
    cols = list(map(list, zip(*rows))) # Transpose
    return (rows, cols)

def scratch(num, rows):
    return list(map(lambda row: list(filter(lambda x: x != num, row)), rows))

if __name__ == "__main__":
    sections = open("input/day04", "r").read().strip().split('\n\n')
    boards = list(map(parse_board, sections[1:]))
    scores = []

    for num in map(int, sections[0].split(',')):
        for i in range(len(boards)):
            rows, cols = boards[i]
            rows = scratch(num, rows)
            cols = scratch(num, cols)

            if not all(rows) or not all(cols):
                scores.append(num * sum(sum(row) for row in rows))
                boards[i] = None
            else:
                boards[i] = (rows, cols)

        boards = list(filter(None, boards))

    print("Part 1:", scores[0])
    print("Part 2:", scores[-1])
